File: Leet Code Solutions/1458-sort-integers-by-the-number-of-1-bits/1458-sort-integers-by-the-number-of-1-bits.java
class Solution {
    public int[] sortByBits(int[] arr) {
        // Convert the input integer array to Integer objects for sorting
        Integer[] integerArr = Arrays.stream(arr).boxed().toArray(Integer[]::new);

        // Create a custom comparator to sort by bit count and then numerically
        Comparator<Integer> comparator = new BitCountComparator();

        // Sort the array using the custom comparator
        Arrays.sort(integerArr, comparator);

        // Convert the sorted array back to primitive integers
        int[] sortedArr = Arrays.stream(integerArr).mapToInt(Integer::intValue).toArray();

        return sortedArr;
    }
}

class BitCountComparator implements Comparator<Integer> {
    @Override
    public int compare(Integer a, Integer b) {
        // Compare based on the count of set bits (1s) in the binary representation
        int bitCountA = Integer.bitCount(a);
        int bitCountB = Integer.bitCount(b);

        if (bitCountA == bitCountB) {
            // If bit counts are the same, compare numerically
            return a - b;
        } else {
            // Sort by the number of set bits in ascending order
            return bitCountA - bitCountB;
        }
    }
}

File: Leet Code Solutions/1458-sort-integers-by-the-number-of-1-bits/1458-sort-integers-by-the-number-of-1-bits.cpp
class Solution {
public:
    static bool compare(int a, int b) {
        // Calculate the number of set bits (1s) in the binary representation of a and b
        int bitCountA = __builtin_popcount(a);
        int bitCountB = __builtin_popcount(b);

        // If the number of set bits is the same for a and b, compare them numerically
        if (bitCountA == bitCountB) {
            return a < b;
        }
        
        // Sort by the number of set bits in ascending order
        return bitCountA < bitCountB;
    }
    
    vector<int> sortByBits(vector<int>& arr) {
        // Use the compare function to sort the vector
        sort(arr.begin(), arr.end(), compare);
        
        // Return the sorted vector
        return arr;
    }
};